runtime/platform/log.cpp

```diff
@@ -84,20 +120,24 @@ void vlogf(
 
   // Maximum length of a log message.
   static constexpr size_t kMaxLogMessageLength = 256;
-  char buf[kMaxLogMessageLength];
-  size_t len = vsnprintf(buf, kMaxLogMessageLength, format, args);
-  if (len >= kMaxLogMessageLength - 1) {
-    buf[kMaxLogMessageLength - 2] = '$';
-    len = kMaxLogMessageLength - 1;
-  }
-  buf[kMaxLogMessageLength - 1] = 0;
+  char buffer[kMaxLogMessageLength];
+
+  const auto write_count =
+      vsnprintf(buffer, kMaxLogMessageLength, format, args);
+  const size_t used_length = (write_count < 0)
+      ? 0
+      : (write_count >= static_cast<int>(kMaxLogMessageLength)
+             ? kMaxLogMessageLength - 1
+             : static_cast<size_t>(write_count));
+  const auto valid_length = get_valid_utf8_prefix_length(buffer, used_length);
```

Comment:
Supposing that the buffer was valid UTF-8 except for possible truncation, you don't have to scan it from the beginning to detect truncation in the middle of a UTF-8 code point. Looking at https://en.wikipedia.org/wiki/UTF-8#Description :

- If the new last byte doesn't have the high bit set, it's valid.
- If it has the second highest bit set, it's the first byte of a code point; drop it and you're done
- Otherwise, scan backward from the end until you find a first byte of a code point (second highest bit set). Look at how many bytes are in the buffer after it; either you have all of its continuation bytes (nothing to drop) or you don't (drop the partial code point).